test/teststl.cpp

```diff
@@ -4411,6 +4413,79 @@ class TestStl : public TestFixture {
               true);
         ASSERT_EQUALS("", errout.str());
     }
+
+    void checkMutexes() {
+        check("void f() {\n"
+              "    static std::mutex m;\n"
+              "    static std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (error) The lock guard won't unlock until the end of the program which could lead to a deadlock.\n", errout.str());
+
+        check("void f() {\n"
+              "    static std::mutex m;\n"
+              "    std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f() {\n"
+              "    std::mutex m;\n"
+              "    std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (error) The mutex is locked at the same scope as the mutex itself.\n", errout.str());
```

Comment:
somehow we need to elaborate. It is not undefined behavior to put a lock_guard at the same scope as a mutex.

Reply:
In this case the mutex serves no purpose. It should be removed or moved to a higher scope to it can catch multithreaded issues. I can change this to a warning as well.